src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -272,11 +209,112 @@ internal void StartPushingToWorkspaceAndNotifyOfOpenDocuments(IEnumerable<Abstra
         {
             AssertIsForeground();
 
-            using (Dispatcher.CurrentDispatcher.DisableProcessing())
+            // If the solution is closing we shouldn't do anything, because all of our state is
+            // in the process of going away. This can happen if we receive notification that a document has
+            // opened in the middle of the solution close operation.
+            if (_solutionIsClosing)
+            {
+                return;
+            }
+
+            // We need to push these projects and any project dependencies we already know about. Therefore, compute the
+            // transitive closure of the projects that haven't already been pushed, keeping them in appropriate order.
+            var visited = new HashSet<AbstractProject>();
+            var inOrderToPush = new List<AbstractProject>();
+
+            void addToInOrderToPush(AbstractProject project)
             {
-                foreach (var hostState in _workspaceHosts)
+                Contract.ThrowIfFalse(ContainsProject(project));
+
+                // Bail out if any of the following is true:
+                //  1. We have already started pushing changes for this project OR
+                //  2. We have already visited this project in a prior recursive call                
+                if (_pushedProjects.Contains(project) || !visited.Add(project))
+                {
+                    return;
+                }
+
+                foreach (var projectReference in project.GetCurrentProjectReferences())
                 {
-                    hostState.StartPushingToWorkspaceAndNotifyOfOpenDocuments(projects);
+                    addToInOrderToPush(GetProject(projectReference.ProjectId));
+                }
+
+                inOrderToPush.Add(project);
+            }
+
+            foreach (var project in projects)
+            {
+                addToInOrderToPush(project);
+            }
+
+            var projectInfos = inOrderToPush.Select(p => p.CreateProjectInfoForCurrentState()).ToImmutableArray();
+
+            // We need to enable projects to start pushing changes to the workspace even before we add the solution/project to the host.
+            // This is required because between the point we capture the project info for current state and the point where we start pushing to the workspace,
+            // project system may send new events on the AbstractProject on a background thread, and these won't get pushed over to the workspace hosts as we didn't set the _pushingChangesToWorkspaceHost flag on the AbstractProject.
+            // By invoking StartPushingToWorkspaceHosts upfront, any project state changes on the background thread will enqueue notifications to workspace hosts on foreground scheduled tasks.
```

Comment:
wrap super long comments please.  100 or 120 columns seems good for github.

Reply:
This code is being moved, and deleted in another refactoring or two.